### ISSUE TITLE
GFP equation

### ISSUE THREAD

**Issue by mscheltienne** [vferat/pycrostates#160]
Looking quickly through GFP stuff, this is the equation I find in MNE:

```
np.sqrt((data**2).mean(axis=1))
```

And this is in pycrostates: https://github.com/vferat/pycrostates/blob/ff3da2b02024d5a54d09878d87d14834859d38dd/pycrostates/preprocessing/extract_gfp_peaks.py#L160

@vferat Any opinion here?

**Comment by vferat** [comment]
The usual formula used in MS analysis:

![global-field-power-eqn-300x75](https://github.com/vferat/pycrostates/assets/28844486/54e818bb-4206-4222-9933-3d0dfbc14cb4)

which is equals to`np.std`

From [Numpy documentation](https://numpy.org/doc/stable/reference/generated/numpy.std.html):
_The standard deviation is the square root of the average of the squared deviations from the mean, i.e., `std = sqrt(mean(x))`, where `x = abs(a - a.mean())**2`._


It seems that MNE formula is missing the `a.mean()` component ( which is `0`  in case of average reference). Do you know it average reference is applied before ?

**Comment by mscheltienne** [comment]
No that was from a generic method, not expecting any reference (could actually be applied to any channel type..)

**Comment by mscheltienne** [comment]
All good, it was already patch. Now 'gfp' is an alias for std for EEG sensors, and RMS for MEG sensors.

**Comment by vferat** [comment]
We might also need to use RMS for MEG data in our computations

**Comment by mscheltienne** [comment]
Oh ! Good point

**Comment by vferat** [comment]
As suggested in [taits 2022](https://www.sciencedirect.com/science/article/pii/S1053811922001355?via%3Dihub) we could use RMS for GFP calculation as well as cosine instead of correlations

**Comment by vferat** [comment]
As `RMS` and `std` are equivalent in case of centered data, we can change GFP computation to always use RMS, and ensure that in case of EEG data, the average reference is set.
We can do the same and change `_corr_vectors` and `_distance_matrix` from `pycrostates.utils` to use `_cosine_similarity`

```
def _cosine_similarity(vector1, vector2):
    dot_product = np.dot(vector1, vector2) / (np.linalg.norm(vector1) * np.linalg.norm(vector2))
    return dot_product
```

I'm creating a new branch to keep things clearer.

**Comment by mscheltienne** [comment]
But are we actually enforcing centered data? I don't think average reference is a requirement at the moment, should it?

**Comment by vferat** [comment]
Yes I agree, the current implementation enforce centered data, and so everything words well for EEG.

However, when [tait 2022](https://www.sciencedirect.com/science/article/pii/S1053811922001355) extend the definition of EEG microstates to MEG microstates, they propose to use RMS as a measure of global field power and cosine similarity as a measure on similarity. This measures extend the current methodology to reference free recordings such a MEG.

In case of centered data, **std = rms** and  **cosine similarity == correlation** , so the pipelines hold in

**Comment by vferat** [comment]
Fixed in [#197](https://github.com/vferat/pycrostates//pull/197)
